cpp/include/wholememory/nvshmem_template.cuh

```diff
@@ -0,0 +1,249 @@
+  ret.w = nvshmem_get<float>(reinterpret_cast<const float*>(source) + 3, pe);
+
+  return ret;
+};
+
+template <typename DataType>
+__inline__ __device__ __host__ void nvshmem_put(DataType* dest, const DataType val, int pe)
+{
+  return;
+};
+
+template <>
+__inline__ __device__ __host__ void nvshmem_put<float>(float* dest, const float val, int pe)
+{
+  return nvshmem_float_p(dest, val, pe);
+}
+template <>
+__inline__ __device__ __host__ void nvshmem_put<double>(double* dest, const double val, int pe)
+{
+  return nvshmem_double_p(dest, val, pe);
+}
+
+template <>
+__inline__ __device__ __host__ void nvshmem_put<int8_t>(int8_t* dest, const int8_t val, int pe)
+{
+  const char& val_int = reinterpret_cast<const char&>(val);
+  return nvshmem_char_p(reinterpret_cast<char*>(dest), val_int, pe);
+}
+
+template <>
+__inline__ __device__ __host__ void nvshmem_put<int16_t>(int16_t* dest, const int16_t val, int pe)
+{
+  return nvshmem_int16_p(dest, val, pe);
+}
+template <>
+__inline__ __device__ __host__ void nvshmem_put<int32_t>(int32_t* dest, const int32_t val, int pe)
+{
+  return nvshmem_int32_p(dest, val, pe);
+}
+template <>
+__inline__ __device__ __host__ void nvshmem_put<int64_t>(int64_t* dest, const int64_t val, int pe)
+{
+  return nvshmem_int64_p(dest, val, pe);
+}
+template <>
+__inline__ __device__ __host__ void nvshmem_put<uint8_t>(uint8_t* dest, const uint8_t val, int pe)
+{
+  return nvshmem_uint8_p(dest, val, pe);
+}
+template <>
+__inline__ __device__ __host__ void nvshmem_put<uint16_t>(uint16_t* dest,
+                                                          const uint16_t val,
+                                                          int pe)
+{
+  return nvshmem_uint16_p(dest, val, pe);
+}
+template <>
+__inline__ __device__ __host__ void nvshmem_put<uint32_t>(uint32_t* dest,
+                                                          const uint32_t val,
+                                                          int pe)
+{
+  return nvshmem_uint32_p(dest, val, pe);
+}
+template <>
+__inline__ __device__ __host__ void nvshmem_put<uint64_t>(uint64_t* dest,
+                                                          const uint64_t val,
+                                                          int pe)
+{
+  return nvshmem_uint64_p(dest, val, pe);
+}
+
+template <>
+__inline__ __device__ __host__ void nvshmem_put<half>(half* dest, const half val, int pe)
+{
+  const int16_t& val_int = reinterpret_cast<const int16_t&>(val);
+  return nvshmem_int16_p(reinterpret_cast<int16_t*>(dest), val_int, pe);
+}
+
+template <>
+__inline__ __device__ __host__ void nvshmem_put<__nv_bfloat16>(__nv_bfloat16* dest,
+                                                               const __nv_bfloat16 val,
+                                                               int pe)
+{
+  const int16_t& val_int = reinterpret_cast<const int16_t&>(val);
+  return nvshmem_int16_p(reinterpret_cast<int16_t*>(dest), val_int, pe);
+}
```

Comment:
If I am not mistaken, it seem like we have more specializations for `nvshmem_get` than `nvshmem_put`, any reason for this? And it will be more readable if we match the order of `nvshmem_get` and `nvshmem_put` specializations; e.g. `nvshmem_get` specializations start with int8_t but `nvshmem_put` specializations start with float.